source/common/common/utility.cc

```diff
@@ -250,10 +250,40 @@ std::string StringUtil::escape(const std::string& source) {
 std::string AccessLogDateTimeFormatter::fromTime(const SystemTime& time) {
   static DateFormatter default_date_format("%Y-%m-%dT%H:%M:%S");
 
-  return fmt::format(
-      "{}.{:03d}Z", default_date_format.fromTime(time),
-      std::chrono::duration_cast<std::chrono::milliseconds>(time.time_since_epoch()).count() %
-          1000);
+  struct CachedTime {
+    std::chrono::seconds epoch_time_seconds;
+    std::string formatted_time;
+    bool initialized_{false};
+  };
+  static thread_local CachedTime cached_time;
```

Comment:
I wonder if we want to use Envoy TLS here. I think what is there right now is a violation of the style guide, since this is a static non-PoD and hence subject to static initialization fiasco.

Reply:
Also, you might want to have the ability to mock this out in tests.

Reply:
I still need to fix this part. I'll follow up on it tomorrow.

Reply:
I was going to mention this before, but this change is somewhat similar to https://github.com/envoyproxy/envoy/blob/master/source/common/http/date_provider_impl.h#L29. You could possibly build on that, though I'm not sure it's worth it. Up to you.

Reply:
Is there an existing API I can use to create a thread-local object and ensure it's destroyed upon thread exit, without also having to publish values to the object from a main thread?

Reply:
I can't speak to the ASAN checker, but the tcmalloc heap checker which we also use has an explicit mode which ignores static allocations (basically only tracks one main starts running).

Reply:
Thinking some more about this, the more I'm concerned that construct on first use (at least in the form recommended by the link in the Envoy style guide) is an antipattern for thread-local objects. The amount of leaked memory is proportional to the number of threads created during the lifetime of the process, so memory will grow without bound if any current or future code tries to does something otherwise reasonable like creating a new thread pool upon config reload.

Reply:
Since the underlying date/time formatting implementation uses `strftime` and thus needs to operate on a character array (rather than a std::string) anyway, I'm just going to concede defeat, replace the cached string with a cached character array, and thus turn the static non-PoD into a static PoD.